### ISSUE TITLE
RFE: Fedora / RHEL / CentOS package

### ISSUE THREAD

**Issue by dralley** [dundee/gdu#57]

**Comment by daniejstriata** [comment]
I created a preliminary SPEC that will create a RPM that will install the precompiled package. I'll see if I can get a rpm that will build the source instead.
[gdu.txt](https://github.com/dundee/gdu/files/6858134/gdu.txt)

I make a tar.gz with the following name:
`gdu_linux_amd64-5.3.0.tar.gz`
And it contains a directory:
```
 gdu-5.3.0/
 gdu-5.3.0/gdu
```
I'll let you know if I get a working RPM that will build from Source instead.

**Comment by daniejstriata** [comment]
Compiling the source fails because I am not pointing to the Go files as seen below:

Package name entry:
Name:           gdu

```
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.i3M1EY
+ umask 022
+ cd /home/danie.dejager/rpmbuild/BUILD
+ cd gdu-5.3.0
+ go build -trimpath -buildmode=pie -mod=readonly -modcacherw -ldflags '-linkmode external -extldflags "" ' -o gdu
no Go files in /home/builduser/rpmbuild/BUILD/gdu-5.3.0
error: Bad exit status from /var/tmp/rpm-tmp.i3M1EY (%build)
```
This is the build command I use:

```
go build \
    -trimpath \
    -buildmode=pie \
    -mod=readonly \
    -modcacherw \
    -o %{name}

```
Contents of source after extracted:
```
build
cmd
codecov.yml
gdu.1
gdu.1.md
go.mod
go.sum
internal
LICENSE.md
Makefile
pkg
README.md
report
snapcraft.yaml
stdout
tui
```

**Comment by dundee** [comment]
Thanks for diving into this!

This simple build command should work:

```
go build \
    -trimpath \
    -buildmode=pie \
    -mod=readonly \
    -modcacherw \
    -o %{name}
    github.com/dundee/gdu/v5/cmd/gdu
```

**Comment by dundee** [comment]
It would be great to also add ldflags and disable CGO:

```
GO111MODULE=on CGO_ENABLED=0 go build \
           -trimpath \
            -buildmode=pie \
            -mod=readonly \
            -modcacherw \
            -ldflags \
              "-s -w \
              -X 'github.com/dundee/gdu/v5/build.Version=$(git describe)' \
              -X 'github.com/dundee/gdu/v5/build.User=$(id -u -n)' \
              -X 'github.com/dundee/gdu/v5/build.Time=$(LC_ALL=en_US.UTF-8 date)'" \
            -o %{name} \
            github.com/dundee/gdu/v5/cmd/gdu
```

**Comment by daniejstriata** [comment]
@dundee Do you require go 1.16 for building at minimum? Neither EL7 or EL8 comes with it standard. They're limited to 1.15 if using standard repos.  Trying to build with go 1.15 failed with this:
`internal/common/ui.go:4:2: package io/fs is not in GOROOT (/usr/lib/golang/src/io/fs)`

I had no issues when I removed version 1.15 installed from EPEL and manually installed go 1.16 and tried with that instead.

I've created a SPEC that installs on EL7 and EL8. You will need to have go 1.16 installed and in path. [gdu_5.3.0-2.spec.txt](https://github.com/dundee/gdu/files/6865274/gdu_5.3.0-2.spec.txt)

To build on EL7 you'll also need to install a newer version of git.

pandoc is required and for EL8 found in powertools repo.

```
→ readelf -h /usr/bin/gdu
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00 
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              DYN (Shared object file)
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x46dbe0
  Start of program headers:          64 (bytes into file)
  Start of section headers:          6055448 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         12
  Size of section headers:           64 (bytes)
  Number of section headers:         28
  Section header string table index: 27
```

**Comment by daniejstriata** [comment]
[gdu.zip](https://github.com/dundee/gdu/files/6865392/gdu.zip)
RPM and RPM SRC

**Comment by dundee** [comment]
Yes, go 1.16 is required, I use some incompatible changes introduced in this version.

Could we setup automatic building somewhere? For example on https://cbs.centos.org/koji/?

**Comment by daniejstriata** [comment]
It won't be possible. There is not a go 1.16 available for RHEL 7 or 8 in EPEL , you will not be able to achieve your goal. EPEL packages may only be built against specific RH channels + EPEL. I'm trying COPR but not sure if that will work.

**Comment by dundee** [comment]
Ok, if it not works then we can probably wait until RHEL 9.

**Comment by dralley** [comment]
RHEL 9 is built from Fedora 34, which has 1.16

**Comment by daniejstriata** [comment]
Can you publish the RPM above with your sources?

**Comment by dundee** [comment]
> Can you publish the RPM above with your sources for now? I'll build a RPM when you release. This RPM works on EL6, 7 and 8.

Yes, added to 5.3.0 release - https://github.com/dundee/gdu/releases/tag/v5.3.0

Could you build for 5.4.0 as well?

**Comment by daniejstriata** [comment]
[gdu-5.4.0.zip](https://github.com/dundee/gdu/files/6899819/gdu-5.4.0.zip)
Here you go.

**Comment by dundee** [comment]
Thanks! Added to release files.

**Comment by daniejstriata** [comment]
@dundee 
Here it is:
[gdu-5.5.0.zip](https://github.com/dundee/gdu/files/6914744/gdu-5.5.0.zip)

**Comment by dundee** [comment]
> @dundee
> Here it is:
> [gdu-5.5.0.zip](https://github.com/dundee/gdu/files/6914744/gdu-5.5.0.zip)

Thank you! Added to release.

**Comment by daniejstriata** [comment]
@dundee Here you go. BTW is the man file up-to-date? It is from Jan 2021.

[gdu5.6.0.zip](https://github.com/dundee/gdu/files/7029478/gdu5.6.0.zip)

**Comment by daniejstriata** [comment]
@dundee Here you go. BTW is the man file up-to-date? It is from Jan 2021.
[gdu-5.6.0.zip](https://github.com/dundee/gdu/files/7029529/gdu-5.6.0.zip)

**Comment by dundee** [comment]
> @dundee Here you go. BTW is the man file up-to-date? It is from Jan 2021.
> [gdu-5.6.0.zip](https://github.com/dundee/gdu/files/7029529/gdu-5.6.0.zip)

Thanks! Uploaded.

The man file was last changed on 20th Jul according to git history.

**Comment by daniejstriata** [comment]
@dundee The date in the file stills says Jan 2021.

**Comment by dundee** [comment]
> @dundee The date in the file stills says Jan 2021.

Ah, you are right, I have missed this one.

**Comment by dundee** [comment]
It will be fixed in next release.

**Comment by daniejstriata** [comment]
@dundee  Here is 5.6.2
[gdu-5.6.2-1.zip](https://github.com/dundee/gdu/files/7071295/gdu-5.6.2-1.zip)

**Comment by dundee** [comment]
Thanks!

**Comment by daniejstriata** [comment]
@dundee Here is 5.7.0
[gdu-5.7.0-1.zip](https://github.com/dundee/gdu/files/7118816/gdu-5.7.0-1.zip)

**Comment by dundee** [comment]
Thanks!

**Comment by daniejstriata** [comment]
@dundee  Here is 5.8.0
[gdu-5.8.0-1.zip](https://github.com/dundee/gdu/files/7217789/gdu-5.8.0-1.zip)
0

**Comment by dundee** [comment]
Thanks!

**Comment by daniejstriata** [comment]
@dundee Please pull the 5.8.0-1 version. There is something wrong with the build. I realised that no matter how it ends up version 5.7

https://github.com/dundee/gdu/v5/cmd/gdu does not resolve to a page anymore. Could that be the reason?

**Comment by dundee** [comment]
Ok, deleted.

The src package looks good to me, but the binary indeed contains version 5.7. Not sure why.

**Comment by dundee** [comment]
> https://github.com/dundee/gdu/v5/cmd/gdu does not resolve to a page anymore. Could that be the reason?

I don't think so. This was never a page, only a module name.

**Comment by daniejstriata** [comment]
@dundee this is the compile command

```
GO111MODULE=on CGO_ENABLED=0 go build \
-trimpath \
-buildmode=pie \
-mod=readonly \
-modcacherw \
-ldflags \
"-s -w \
-X 'github.com/dundee/gdu/v5/build.Version=$(git describe)' \
-X 'github.com/dundee/gdu/v5/build.User=$(id -u -n)' \
-X 'github.com/dundee/gdu/v5/build.Time=$(LC_ALL=en_US.UTF-8 date)'" \
-o %{name} github.com/dundee/gdu/v5/cmd/gdu


```

**Comment by dundee** [comment]
Oh, I see the problem. The 5.8.0 tag is not annotated. Therefore `git describe` will not show it.

I will have to release 5.8.1

**Comment by daniejstriata** [comment]
I just saw the same thing. Thanks @dundee

**Comment by daniejstriata** [comment]
@dundee I notice that pandoc complains 
```pandoc: Could not parse YAML header: UnexpectedEvent {_received = Just (EventMappingStart Nothing), _expected = Nothing}```

Here is the 5.8.1 rpm
[gdu-5.8.1.zip](https://github.com/dundee/gdu/files/7230779/gdu-5.8.1.zip)
m

**Comment by dundee** [comment]
There is now variable for date in the markdown source for man page: `date: {{date}}`

I would suggest to use `gdu.1`, which is contained in the archive a not run pandoc (also remove it from build dependencies as well) at all. I have done the same in Arch package - https://aur.archlinux.org/cgit/aur.git/commit/?h=gdu&id=e3e0b7c48609200413ea88c5c3384bf70a25c803

Pandoc is huge and I am running it anyway for every release localy.

**Comment by daniejstriata** [comment]
@dundee That's been rectified.

Here you go. 5.8.1-2
[gdu-5.8.1-2.zip](https://github.com/dundee/gdu/files/7238820/gdu-5.8.1-2.zip)

**Comment by dundee** [comment]
Thanks, added.

**Comment by dundee** [comment]
> It won't be possible. There is not a go 1.16 available for RHEL 7 or 8 in EPEL , you will not be able to achieve your goal. EPEL packages may only be built against specific RH channels + EPEL. I'm trying COPR but not sure if that will work.

Do you think we can package for [EPEL Next](https://docs.fedoraproject.org/en-US/epel/epel-about-next/)?

**Comment by daniejstriata** [comment]
@dundee here is 5.10.0.
[gdu-5.10.0-1.zip](https://github.com/dundee/gdu/files/7509953/gdu-5.10.0-1.zip)

**Comment by dundee** [comment]
Thanks!

**Comment by daniejstriata** [comment]
@dundee Here is 5.10.1
[gdu-5.10.1-1.zip](https://github.com/dundee/gdu/files/7590737/gdu-5.10.1-1.zip)

**Comment by dundee** [comment]
Thank you!

**Comment by daniejstriata** [comment]
@dundee Here is 5.11.0
[gdu-5.11.0-1.zip](https://github.com/dundee/gdu/files/7613571/gdu-5.11.0-1.zip)

**Comment by dundee** [comment]
Thanks!

**Comment by daniejstriata** [comment]
@dundee I created a new  package for 5.11.0 compiled with go 1.17.4 
[gdu-5.11.0-2.zip](https://github.com/dundee/gdu/files/7648875/gdu-5.11.0-2.zip)
1

**Comment by dundee** [comment]
@daniejstriata Great! Uploaded.

**Comment by daniejstriata** [comment]
@dundee Here is 5.12.0
[gdu-5.12.0-1.zip](https://github.com/dundee/gdu/files/7652463/gdu-5.12.0-1.zip)

**Comment by dundee** [comment]
Hi @daniejstriata, could you pack 5.12.1?

**Comment by daniejstriata** [comment]
@dundee Here you go. 5.12.1
[gdu-5.12.1-1.zip](https://github.com/dundee/gdu/files/7709548/gdu-5.12.1-1.zip)

**Comment by dundee** [comment]
Thanks!

**Comment by daniejstriata** [comment]
@dundee Here is 5.13.0
[gdu-5.13.0-1.zip](https://github.com/dundee/gdu/files/7964467/gdu-5.13.0-1.zip)

**Comment by dundee** [comment]
Thank you!

**Comment by daniejstriata** [comment]
@dundee Here is 5.13.1
[gdu-5.13.1-1.zip](https://github.com/dundee/gdu/files/8038199/gdu-5.13.1-1.zip)

**Comment by dundee** [comment]
Added, thanks:!

**Comment by daniejstriata** [comment]
@dundee  Here is 5.13.2
[gdu-5.13.2-1.zip](https://github.com/dundee/gdu/files/8107543/gdu-5.13.2-1.zip)

**Comment by dundee** [comment]
Thanks!

**Comment by daniejstriata** [comment]
@dundee Here is 5.14.0. Compiled with go 1.17.10. To add this in to EPEL gdu would need to build against go 1.17.5
[gdu-5.14.0-1.zip](https://github.com/dundee/gdu/files/8777660/gdu-5.14.0-1.zip)

**Comment by daniejstriata** [comment]
@dundee Not making progress adding gdu to Fedora. Will try to submit it again.

here is the latest RPM
[gdu-5.15.0-1.x86_64.zip](https://github.com/dundee/gdu/files/9299945/gdu-5.15.0-1.x86_64.zip)

**Comment by daniejstriata** [comment]
@dundee Here is the RPM
[gdu-5.16.0-1.zip](https://github.com/dundee/gdu/files/9532775/gdu-5.16.0-1.zip)

**Comment by dundee** [comment]
Thank you!

**Comment by daniejstriata** [comment]
@dundee Here is 5.17.1
[gdu-5.17.1-1.zip](https://github.com/dundee/gdu/files/9555214/gdu-5.17.1-1.zip)

**Comment by dundee** [comment]
Thanks!

**Comment by daniejstriata** [comment]
@dundee Here is 5.18.1

**Comment by daniejstriata** [comment]
@dundee Here is 5.19.0
[gdu-5.19.0-1.x86_64.zip](https://github.com/dundee/gdu/files/9661850/gdu-5.19.0-1.x86_64.zip)

**Comment by dundee** [comment]
Thanks!

**Comment by daniejstriata** [comment]
@dundee Here is 5.20.0 built using go 1.18.7 for `x86_64` and `aarch64` going forward.
[gdu-5.20.0.zip](https://github.com/dundee/gdu/files/9855487/gdu-5.20.0.zip)

**Comment by dundee** [comment]
Thanks!

**Comment by daniejstriata** [comment]
@dundee Here is 5.21.0
[gdu-5.21.0-1.x86_64.zip](https://github.com/dundee/gdu/files/10347501/gdu-5.21.0-1.x86_64.zip)
[gdu-5.21.0-1.aarch64.zip](https://github.com/dundee/gdu/files/10347602/gdu-5.21.0-1.aarch64.zip)

**Comment by dundee** [comment]
Thank you! Uploaded.

**Comment by dundee** [comment]
I have also sent you invitation for collaboration so you can uplad the artifacts yourself.

**Comment by dralley** [comment]
An option you could consider is to set up a COPR, and have CI trigger a COPR build.  That would also provide users with the ability to get automatic updates via their package manager.

**Comment by daniejstriata** [comment]
I'll look into setting up a COPR repo

**Comment by xmstspider** [comment]
> @dundee Here is 5.21.0 [gdu-5.21.0-1.x86_64.zip](https://github.com/dundee/gdu/files/10347501/gdu-5.21.0-1.x86_64.zip) [gdu-5.21.0-1.aarch64.zip](https://github.com/dundee/gdu/files/10347602/gdu-5.21.0-1.aarch64.zip)

I tried to build latest gdu (5.21.1) using slightly modified spec file from your 5.20 version, and my Fedora 36 is throwing the following error:
```
RPM build warnings:
    Missing build-id in /home/user/rpmbuild/BUILDROOT/gdu-5.21.1-1.x86_64/usr/bin/gdu

```

The solution is to change the go build line in the spec by adding `-linkmode=external` to the argument list (see below):

```-ldflags \
"-linkmode=external \
-s -w \
-X 'github.com/dundee/gdu/v5/build.Version=$(git describe)' \
-X 'github.com/dundee/gdu/v5/build.User=$(id -u -n)' \
-X 'github.com/dundee/gdu/v5/build.Time=$(LC_ALL=en_US.UTF-8 date)'" \
-o %{name} github.com/dundee/gdu/v5/cmd/gdu
```

May be worth checking if it possibly affects RHEL builds...

**Comment by daniejstriata** [comment]
@xmstspider I have a new SPEC file that supports builds from Fedora.

**Comment by souvlakeeb** [comment]
Bump.

**Comment by daniejstriata** [comment]
The latest Fedora 41 build fails when setting linkmode=external. I've built all the gdu packages in copr without this modification on all Fedora and RHEL based distros without issues.
https://copr.fedorainfracloud.org/coprs/faramirza/gdu/

**Comment by danie-dejager** [comment]
I'm happy to see that gdu is in Fedora. 
https://src.fedoraproject.org/rpms/gdu
